### PR TITLE
Support CPU inference with VSX PowerPC ISA

### PR DESCRIPTION
**Pull request by ChipKerchner**
Support CPU inference with VSX PowerPC ISA.

This PR provides basic CPU support for a PowerPC (Power9 & POWER10).  Additional PRs in the future will address further performance.

To build:

```
docker build --format docker -t vllm:v0.5.0-cpu -f Dockerfile.ppc64le .
```

Run the api server with:

```
docker run -ti --rm   -v ~/.cache/huggingface:/root/.cache/huggingface   --env "HUGGING_FACE_HUB_TOKEN=<TOKEN>"   -p 8000:8000   --ipc=host vllm:v0.5.0-cpu  --model=facebook/opt-125m   --dtype=bfloat16
```

or

```
docker run -ti --rm   -v ~/.cache/huggingface:/root/.cache/huggingface   --env "HUGGING_FACE_HUB_TOKEN=<TOKEN>"   -p 8000:8000   --ipc=host vllm:v0.5.0-cpu  --model=TinyLlama/TinyLlama-1.1B-Chat-v1.0   --dtype=float32
```

Testing has been done on a Power10 with the following models:

```
models--facebook--opt-125m
models--ibm-granite--granite-7b-base
models--openai-community--gpt2
models--TinyLlama--TinyLlama-1.1B-Chat-v1.0
```

### CODE REVIEW

**Comment by ChipKerchner** [comment]
Is there a way to have ruff ignore these?  Ruff is not correct.

```
Run codespell --toml pyproject.toml
./csrc/cpu/cpu_types_vsx.hpp:352: ans ==> and
```

**Comment by mgoin** [comment]
Hi @ChipKerchner, there is a section for codespell in the pyproject.toml where you can set special cases:
https://github.com/vllm-project/vllm/blob/07feecde1a69859d565786a7ad64c0f604f17b28/pyproject.toml#L61-L63

**Comment by ChipKerchner** [comment]
Would it be better to add an ignore the word `ans` or to add a new file location `csrc/cpu`?

File: csrc/ops.h [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <optional>
```
Comment (mgoin) [inline csrc/ops.h:3]: Why is this needed?

File: Dockerfile.ppc64le [hunk 2]
```diff
@@ -0,0 +1,19 @@
+FROM mambaorg/micromamba
+ARG MAMBA_DOCKERFILE_ACTIVATE=1
+USER root
+
+RUN apt-get update  -y     && apt-get install -y git wget vim numactl gcc-12 g++-12 protobuf-compiler libprotobuf-dev     && update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-12 10 --slave /usr/bin/g++ g++ /usr/bin/g++-12
+
+RUN micromamba install -y -n base -c rocketce  -c defaults     python=3.10     pytorch-cpu     numba     llvmlite     httptools &&     micromamba clean --all --yes
```
Comment (mgoin) [inline Dockerfile.ppc64le:7]: Can you add a note that pytorch is installed here, rather than in the requirements-cpu.txt. How is the versioning controlled? It seems to be unrestricted here atm

**Review by mgoin — COMMENTED** [review]
Thanks for the new architecture contribution! The implementation looks clean, I just have some questions about the dependencies and build changes

File: csrc/ops.h (hunk 1 above)
Comment (ChipKerchner) [reply csrc/ops.h:3]: Without this I get these errors:

```
In file included from /workspace/vllm/csrc/cpu/torch_bindings.cpp:2:
/workspace/vllm/csrc/ops.h:58:36: error: ‘optional’ in namespace ‘std’ does not name a template type
   58 |                         const std::optional<torch::Tensor>& bias);
      |                                    ^~~~~~~~
/workspace/vllm/csrc/ops.h:1:1: note: ‘std::optional’ is defined in header ‘<optional>’; did you forget to ‘#include <optional>’?
  +++ |+#include <optional>
```

Comment (mgoin) [reply csrc/ops.h:3]: Yes, I believe c++17 is enabled by default when using the CUDA backend because it is required for CUTLASS/modern CUDA. This is fine then

File: cmake/cpu_extension.cmake [hunk 3]
```diff
@@ -68,8 +70,15 @@ if (AVX512_FOUND AND NOT AVX512_DISABLED)
 elseif (AVX2_FOUND)
     list(APPEND CXX_COMPILE_FLAGS "-mavx2")
     message(WARNING "vLLM CPU backend using AVX2 ISA")
+elseif (${CMAKE_SYSTEM_PROCESSOR} MATCHES "ppc64")
+    message(STATUS "PowerPC detected")
+    # Check for PowerPC VSX support
+    list(APPEND CXX_COMPILE_FLAGS
+        "-mvsx"
+        "-mcpu=native"
+        "-mtune=native")
```
[the remark this answers is not in the document]
Comment (ChipKerchner) [reply cmake/cpu_extension.cmake:79]: The minimum gcc version would be 10.3 but since we are using gcc-12 are checks really necessary?

File: Dockerfile.ppc64le (hunk 2 above)
Comment (ChipKerchner) [reply Dockerfile.ppc64le:7]: Right now the available packages are limited for PowerPC (aka pytorch and triton).  Until the versions required in `requirements-cpu.txt` are met for PPC, this might be the best that can be done.

[the remark this answers is not in the document]
Comment (mgiessing) [reply Dockerfile.ppc64le:7]: Hi @mgoin the reason for using mamba/conda based builds is that IBM provides optimized packages for ppc64le processors by the [open-ce project](https://github.com/open-ce/open-ce) which are then built and hosted by universities like [Oregon State University](https://ftp.osuosl.org/pub/open-ce/1.11.0-p10/) or partners like [RocketCE](https://anaconda.org/rocketce).

The issue with venv & pip is that many open-source project do not provide pre-built wheels for Power architecture and thus using `pip install` often fails on non-x86/aarch64 platforms

[the remark this answers is not in the document]
Comment (mgoin) [reply Dockerfile.ppc64le:7]: Appreciate the clarity! Could you please add a comment in the dockerfile clarifying this? It would be useful to document this for users interested in understanding or attempting to build without docker.

File: cmake/cpu_extension.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgoin) [reply cmake/cpu_extension.cmake:79]: Sure that is reasonable. I am just trying to match this implementation up with the standard set by previous CPU contributions. Specifying details like this are useful when attempting to build without docker

**Review by mgoin — APPROVED** [review]
Looks good to me pending green CI, thanks!

**Comment by jwoehr** [comment]
Trying to build this but it ends like this:
```
    Finished `release` profile [optimized] target(s) in 1m 52s
📦 Built wheel for abi3 Python ≥ 3.7 to /xet-core/hf_xet/target/wheels/hf_xet-1.1.5-cp37-abi3-linux_ppc64le.whl
/xet-core/hf_xet/target/wheels/hf_xet-1.1.5-cp37-abi3-linux_ppc64le.whl
Successfully built /hf_wheels/hf_xet-1.1.5-cp37-abi3-linux_ppc64le.whl
--> 7d3c88257d76
[8/10] STEP 9/9: RUN --mount=type=cache,target=/root/.cache/uv     --mount=type=bind,from=torch-builder,source=/torchwheels/,target=/torchwheels/,ro     --mount=type=bind,from=arrow-builder,source=/arrowwheels/,target=/arrowwheels/,ro     --mount=type=bind,from=cv-builder,source=/opencvwheels/,target=/opencvwheels/,ro     --mount=type=bind,from=numa-builder,source=/numactl/,target=/numactl/,rw     --mount=type=bind,src=.,dst=/src/,rw     source /opt/rh/gcc-toolset-13/enable &&     uv pip install /opencvwheels/*.whl /arrowwheels/*.whl /torchwheels/*.whl &&     sed -i -e 's/.*torch.*//g' /src/pyproject.toml /src/requirements/*.txt &&     uv pip install pandas pythran pybind11 /hf_wheels/*.whl &&     make -C /numactl install &&     export PKG_CONFIG_PATH=$(find / -type d -name "pkgconfig" 2>/dev/null | tr '\n' ':') &&     uv pip install -r /src/requirements/common.txt -r /src/requirements/cpu.txt -r /src/requirements/build.txt --no-build-isolation &&     cd /src/ &&     uv build --wheel --out-dir /vllmwheel/ --no-build-isolation &&     uv pip install /vllmwheel/*.whl
Using Python 3.12.9 environment at: /opt/vllm
Resolved 16 packages in 100ms
   Building pillow==11.2.1
      Built pillow==11.2.1
Prepared 6 packages in 23.91s
Installed 15 packages in 717ms
 + filelock==3.18.0
 + fsspec==2025.5.1
 + jinja2==3.1.6
 + markupsafe==3.0.2
 + mpmath==1.3.0
 + networkx==3.5
 + numpy==2.3.1
 + opencv-python-headless==4.11.0.86 (from file:///opencvwheels/opencv_python_headless-4.11.0.86-cp312-cp312-linux_ppc64le.whl)
 + pillow==11.2.1
 + pyarrow==19.0.1 (from file:///arrowwheels/pyarrow-19.0.1-cp312-cp312-linux_ppc64le.whl)
 + sympy==1.13.1
 + torch==2.6.0 (from file:///torchwheels/torch-2.6.0-cp312-cp312-linux_ppc64le.whl)
 + torchaudio==2.6.0 (from file:///torchwheels/torchaudio-2.6.0-cp312-cp312-linux_ppc64le.whl)
 + torchvision==0.21.0 (from file:///torchwheels/torchvision-0.21.0-cp312-cp312-linux_ppc64le.whl)
 + typing-extensions==4.14.0
sed: can't read /src/pyproject.toml: No such file or directory
sed: can't read /src/requirements/*.txt: No such file or directory
Error: building at STEP "RUN --mount=type=cache,target=/root/.cache/uv --mount=type=bind,from=torch-builder,source=/torchwheels/,target=/torchwheels/,ro --mount=type=bind,from=arrow-builder,source=/arrowwheels/,target=/arrowwheels/,ro --mount=type=bind,from=cv-builder,source=/opencvwheels/,target=/opencvwheels/,ro --mount=type=bind,from=numa-builder,source=/numactl/,target=/numactl/,rw --mount=type=bind,src=.,dst=/src/,rw source /opt/rh/gcc-toolset-13/enable &&     uv pip install /opencvwheels/*.whl /arrowwheels/*.whl /torchwheels/*.whl &&     sed -i -e 's/.*torch.*//g' /src/pyproject.toml /src/requirements/*.txt &&     uv pip install pandas pythran pybind11 /hf_wheels/*.whl &&     make -C /numactl install &&     export PKG_CONFIG_PATH=$(find / -type d -name "pkgconfig" 2>/dev/null | tr '\n' ':') &&     uv pip install -r /src/requirements/common.txt -r /src/requirements/cpu.txt -r /src/requirements/build.txt --no-build-isolation &&     cd /src/ &&     uv build --wheel --out-dir /vllmwheel/ --no-build-isolation &&     uv pip install /vllmwheel/*.whl": while running runtime: exit status 2
``
